### ISSUE TITLE
Infinite memory leak on folly string header during compilation with GCC 6.1.1

### ISSUE THREAD

**Issue by entd** [facebook/folly#418]
I have trivial folly application: https://github.com/entd/follyKiller
When I compile with folly/String.h include it just eats entire ram on my computer until OOM kicks in.
Is it folly problem, or more compiler bug? 

System specification:
GCC 6.1.1:
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc-multilib/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release
Thread model: posix
gcc version 6.1.1 20160501 (GCC) 
```
I'm using Manaro linux, while it is not supported platform it uses unmodified upstream as much as it can.

Project example, which show the bug: https://github.com/entd/follyKiller

I used recent folly which was cloned from github.

**Comment by 2underscores-vic** [comment]
This command:
```
libtool: compile:  g++ -DHAVE_CONFIG_H -I./.. -pthread -I/home/__vic/include -I/usr/local/ssl/include -std=gnu++1y -g -O2 -MT detail/CacheLocality.lo -MD -MP -MF detail/.deps/CacheLocality.Tpo -c detail/CacheLocality.cpp -o detail/CacheLocality.o
```
ate 8Gb of RAM + 20Gb of swap (it's all awailable memory) then g++ died with internal compiler error.
Just tried to build folly (from master) using GCC 6.1

**Comment by kornrunner** [comment]
This can be reproduced on Fedora 24 (gcc 6.1.1, boost 1.60.0)

**Comment by Orvid** [comment]
This is probably a compiler bug.

**Comment by kornrunner** [comment]
Seems so, build passes with `-std=c++0x` flag (vs `c++14` or `c++1y`).

**Comment by entd** [comment]
It would be interesting to find what is happening c++ 14 shouldn't make
much difference from c++ 11 in template handling.
2016 birž. 27 01:27 popiet "Boris Momčilović" <notifications@github.com>
rašė:

**Comment by 0x1997** [comment]
I can confirm. Downgrading gcc to 5.3.0 or setting CXXFLAGS to `-std=c++11` solves the issue.

**Comment by Romain-Geissler-1A** [comment]
Hi,

I have been investigation this issue, and apparently it was already reported to the gcc team with a build issue with facebook/wangle. See here https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70977.

A simple reproducer for this bug is the following (triggered either in C++11 or C++14 with gcc 6.1.1, the reason it only affects Folly in C++14 is because std::max with initializer_list was changed in C++14 (not-constexpr vs constexpr))

    #include <initializer_list>

    template<typename T> constexpr T fake_max(std::initializer_list<T> init_list)
    {
        return *init_list.begin();
    }

    template <typename T> void f() 
    {
        const int a = fake_max({1, 2});
    }

Apparently a fix in gcc is already being worked on.

Cheers,
Romain

**Comment by Romain-Geissler-1A** [comment]
In the meantime, people like me who definitely need to build Folly with C++14 and gcc 6 can apply the following change to avoid the gcc bug: in folly/Conv.h, in the method `template <class Src> typename std::enable_if<std::is_floating_point<Src>::value, size_t>::type estimateSpaceNeeded(Src value)` you can adapt the call to `std::max` with an initializer list this way (note: I am using the master branch, not the release 0.57 from last year):

    std::initializer_list<int> initList = {
        // E.g. 1.1111111111111111E-100.
        kMaxMantissaSpace + kMaxExponentSpace,
        // E.g. 0.000001.1111111111111111, if kConvMaxDecimalInShortestLow is -6.
        kMaxMantissaSpace - detail::kConvMaxDecimalInShortestLow,
        // If kConvMaxDecimalInShortestHigh is 21, then 1e21 is the smallest
        // number > 1 which ToShortest outputs in exponential notation,
        // so 21 is the longest non-exponential number > 1.
        detail::kConvMaxDecimalInShortestHigh
    };  
    static const int kMaxPositiveSpace = std::max(initList);

**Comment by Romain-Geissler-1A** [comment]
FYI: the fix was just backported to the "gcc-6-branch" and gcc 6.2.0 release will happen in the coming weeks. Reference: https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=75e7d66079cd68e5d75147d0c60f2d16f3d09c58

**Comment by Romain-Geissler-1A** [comment]
To be closed now that gcc 6.2.0 has been released.
